File: Good Subtrees[5-May-2023]/Solution.cpp
//  >> C++ CODE <<


class Solution {
public:
    int goodSubtrees(Node* root, int k) {
        
        int ans = 0;
        solve(root, k, ans);
        return ans;
    }
    
private:
    unordered_set<int> solve(Node* root, int k, int& ans) {
        if (root == nullptr) {
            return {};
        }
        
        unordered_set<int> left = solve(root->left, k, ans);
        unordered_set<int> right = solve(root->right, k, ans);
        
        left.insert(root->data);
        for (auto node : right) {
            left.insert(node);
        }
        
        if (left.size() <= k) {
            ans++;
        }
        
        return left;
    }
};

File: Good Subtrees[5-May-2023]/solution.py
#  >> PYTHON CODE <<

class Solution:
    def goodSubtrees(self, root, k):
        def solve(root,k,ans):
            if root is None:
                return set()
            
            left=solve(root.left,k,ans)
            right=solve(root.right,k,ans)
            
            left.add(root.data)
            left.update(right)
            
            if(len(left)<=k):
                ans[0]+=1
            
            return left
            
            
        ans=[0]
        solve(root,k,ans)
        
        return ans[0]